src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11SecretKeyFactory.java

```diff
@@ -52,107 +54,240 @@ final class P11SecretKeyFactory extends SecretKeyFactorySpi {
     // algorithm name
     private final String algorithm;
 
+    // PBEKeyInfo if algorithm is PBE
+    private final PBEKeyInfo svcPbeKi;
+
     P11SecretKeyFactory(Token token, String algorithm) {
         super();
         this.token = token;
         this.algorithm = algorithm;
+        this.svcPbeKi = getPBEKeyInfo(algorithm);
     }
 
-    private static final Map<String,Long> keyTypes;
+    private static final Map<String, KeyInfo> keyInfo = new HashMap<>();
+    private static final KeyInfo HMAC = new KeyInfo("HMAC", PCKK_HMAC);
+    private static final KeyInfo SSLMAC = new KeyInfo("SSLMAC", PCKK_SSLMAC);
+
+    static KeyInfo getKeyInfo(String algo) {
+        KeyInfo ki = keyInfo.get(algo);
+        if (ki == null) {
+            String algoUpper = algo.toUpperCase(Locale.ENGLISH);
+            ki = keyInfo.get(algoUpper);
+            if (ki == null) {
+                if (algoUpper.startsWith("HMAC")) {
+                    return HMAC;
+                } else if (algoUpper.startsWith("SSLMAC")) {
+                    return SSLMAC;
+                }
+            }
+        }
+        return ki;
+    }
+
+    static PBEKeyInfo getPBEKeyInfo(String algo) {
+        if (getKeyInfo(algo) instanceof PBEKeyInfo pbeKi) {
+            return pbeKi;
+        }
+        return null;
+    }
+
+    private static void putKeyInfo(KeyInfo ki) {
+        keyInfo.put(ki.algo, ki);
+        keyInfo.put(ki.algo.toUpperCase(Locale.ENGLISH), ki);
+    }
+
+    static sealed class KeyInfo permits PBEKeyInfo {
+        public final String algo;
+        public final long keyType;
+
+        KeyInfo(String algo, long keyType) {
+            this.algo = algo;
+            this.keyType = keyType;
+        }
+
+        static boolean checkUse(KeyInfo ki, KeyInfo si) {
```

Comment:
More comment on ki and si and how this method is used is needed as it's not obvious when looking at the code here.